src/rabbitmq.js

```diff
@@ -0,0 +1,354 @@
+    this.log.error({ err: err }, 'connection has caused an error')
+    throw err
+  }
+
+  /**
+   * Error handler for the RabbitMQ channel.
+   * @private
+   * @throws Error
+   * @param {object} err Error object from event.
+   */
+  _channelErrorHandler (err: Error) {
+    this.log.error({ err: err }, 'channel has caused an error')
+    throw err
+  }
+
+  /**
+   * Check to see if model is connected.
+   *
+   * @return {Boolean} True if model is connected and channel is established.
+   */
+  _isConnected (): boolean {
+    return !!(this._isPartlyConnected() && this.channel)
+  }
+
+  /**
+   * Check to see if model is _partially_ connected. This means that the
+   * connection was established, but the channel was not.
+   *
+   * @return {Boolean} True if connection is established.
+   */
+  _isPartlyConnected (): boolean {
+    return !!(this.connection)
+  }
+
+  /**
+   * Helper function to consolidate logic for subscribing to queues. Stores
+   * information about what is subscribed and is responsible for asserting
+   * exchanges and queues into existance.
+   *
+   * @private
+   * @param {Object} opts Object describing the exchange connection.
+   * @param {String} opts.exchange Name of exchange.
+   * @param {String} opts.handler Handler of jobs.
+   * @param {String} opts.type Type of exchange: 'fanout' or 'topic'.
+   * @param {String} [opts.routingKey] Routing key for a topic exchange.
+   * @return {Promise} Promise resolved when subcribed to exchange.
+   */
+  _subscribeToExchange (opts: SubscribeObject): Promise {
+    const log = this.log.child({
+      method: '_subscribeToExchange',
+      opts: opts
+    })
+    log.info('subscribing to exchange')
+    if (!this._isConnected()) {
+      return Promise.reject(new Error('must .connect() before subscribing'))
+    }
+    if (opts.type === 'topic' && !opts.routingKey) {
+      return Promise.reject(new Error('routingKey required for topic exchange'))
+    }
+    let subscribedKey = `${opts.type}:::${opts.exchange}`
+    if (opts.type === 'topic') {
+      subscribedKey = `${subscribedKey}:::${opts.routingKey}`
+    }
+    if (this.subscribed.has(subscribedKey)) {
+      log.warn(`already subscribed to ${opts.type} exchange`)
+      return Promise.resolve()
+    }
+    return Promise
+      .resolve(
+        this.channel.assertExchange(
+          opts.exchange,
+          opts.type,
+          { durable: false }
+        )
+      )
+      .then(() => {
+        log.info('exchange asserted')
+        let queueName = `ponos.${opts.exchange}`
+        if (opts.type === 'topic') {
+          queueName = `${queueName}.${opts.routingKey}`
+        }
+        return Promise
+          .resolve(this.channel.assertQueue(queueName, { exclusive: true }))
+      })
+      .then((queueInfo) => {
+        const queue = queueInfo.queue
+        log.info({ queue: queue }, 'queue asserted')
+        log.info('binding queue')
+        if (!opts.routingKey) {
```

Comment:
you enforce routing key above but set it null here? routing key only valid for topic exchange?